File: flink-java-code/src/main/java/com/atguigu/day05/SideOutputExample.java
package com.atguigu.day05;

import com.atguigu.day02.SensorReading;
import com.atguigu.day02.SensorSource;
import org.apache.commons.math3.fitting.leastsquares.EvaluationRmsChecker;
import org.apache.flink.streaming.api.datastream.DataStream;
import org.apache.flink.streaming.api.datastream.DataStreamSource;
import org.apache.flink.streaming.api.datastream.SingleOutputStreamOperator;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.streaming.api.functions.ProcessFunction;
import org.apache.flink.util.Collector;
import org.apache.flink.util.OutputTag;

public class SideOutputExample {
    private static OutputTag<String> sideOutput = new OutputTag<String>("side-output"){};
    public static void main(String[] args) throws Exception {
        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
        env.setParallelism(1);
        DataStream<SensorReading> stream = env.addSource(new SensorSource());
        SingleOutputStreamOperator<SensorReading> waring = stream.process(new FreezingAlarm());
waring.print();
waring.getSideOutput(sideOutput).print();
env.execute();
    }
public static class FreezingAlarm extends ProcessFunction<SensorReading, SensorReading>{
    @Override
    public void processElement(SensorReading sensorReading, Context context, Collector<SensorReading> collector) throws Exception {
        if (sensorReading.temperature < 32.0) {
            context.output(sideOutput, "温度小于32度！");
        }
        collector.collect(sensorReading);
    }
    }
}

File: FlinkTutorial/src/main/scala/com/atguigu/day05/SideOutputExample.scala
package com.atguigu.day05

import com.atguigu.day02.{SensorReading, SensorSource}
import org.apache.flink.streaming.api.functions.ProcessFunction
import org.apache.flink.streaming.api.scala._
import org.apache.flink.util.Collector

object SideOutputExample {
  private val sideOutput = new OutputTag[String]("side-output")
  def main(args: Array[String]): Unit = {
    val env: StreamExecutionEnvironment = StreamExecutionEnvironment.getExecutionEnvironment
    env.setParallelism(1)
    val stream: DataStream[SensorReading] = env.addSource(new SensorSource)
val wearing = stream.process(new FreezingAlarm)
    wearing.print()
    wearing.getSideOutput(sideOutput).print()
    env.execute()
  }
  class FreezingAlarm extends ProcessFunction[SensorReading, SensorReading]{
    override def processElement(i: SensorReading, context: ProcessFunction[SensorReading, SensorReading]#Context, collector: Collector[SensorReading]): Unit = {
      if (i.temaperature < 32.0){
        context.output(sideOutput, "传感器ID为：" + i.id + "的传感器温度小于32度！")
      }
      collector.collect(i)
    }
  }
}
